### ISSUE TITLE
Oracle migrations are not transactional by default

### ISSUE THREAD

**Issue by fairking** [fluentmigrator/fluentmigrator#1262]
**Describe the bug**
If I run migrations on oracle db and after some error happen I clearly see there are partially some tables crated. I cannot see it in mssql.

**To Reproduce**
Create a migrator with some table creation and then add next line which will throw an error. You will see that the first table is created.

**Expected behavior**
It should rollback whole transaction without creating anything.

**Information (please complete the following information):**
 - OS: Windows 10
 - Platform .NET Core 3.1.x
 - FluentMigrator version 3.2.6
 - FluentMigrator runner
 - Database Management System: Oracle
 - Database Management System Version: "Oracle 18xe"

**Additional context**
```
    [Migration(201909232030)]
    public class MIG_20190923_2030_Init : Migration
    {
        public override void Up()
        {
            ...
```

**Comment by jzabroski** [comment]
@fairking Has oracle added transactional DDL?

https://www.google.com/search?q=oracle+transactional+ddl

![image](https://user-images.githubusercontent.com/447485/83785315-228d2200-a65f-11ea-9921-3e71e3fc78d2.png)

**Comment by jzabroski** [comment]
To elaborate, one of the amazing and under-recognized features of SQL Server is transactional DDL.  Very few databases have this feature.  It doesn't always work perfectly, especially with synonyms and how name binding resolution works in SQL Server, but when it does, it saves a ton of time.

**Comment by jzabroski** [comment]
No worries.  Oracle shops tend to use the Down part of migrations a lot :)

**Comment by fairking** [comment]
>> No worries. Oracle shops tend to use the Down part of migrations a lot :)
I have to write tons of down scripts which I don't have. Second thing whic is worrying me is all my transactions in all my requests are going to hell. :-(

**Comment by jzabroski** [comment]
See: http://www.dba-oracle.com/oracle_tips_db_copy.htm

You need an easy way to clone your pre-prod db environment so that you can effectively flashback your whole db schema.

**Comment by fairking** [comment]
I hope I can also use `AutoReversingMigration` for oracle issue.

**Comment by jzabroski** [comment]
Yes, you can.  I am not sure how well supported AutoReversingMigration is.  I personally only use ForwardOnlyMigration.  I work on very huge db projects with 1000 tables/indexes/views and hundreds of stored procedures, and the cost of rolling back vs. testing well is rather no-brainer.

**Comment by fairking** [comment]
> the cost of rolling back vs. testing well is rather no-brainer
I agree. That's why am looking for alternatives. Db backup is the second alternative one.

**Comment by fairking** [comment]
What I would like it to be is to have `AutoReversingMigration` which could allow for Update/Insert/Delete/Execute. Then I could always put a check if the following Update/Insert/Delete/Execute required or not. Something like: `if {row doesn't exist} do {insert}` or `if {row exists} do {delete}` and so on. What do you think @jzabroski ?

**Comment by fairking** [comment]
This is one of the examples how do I solve it :-) :
![image](https://user-images.githubusercontent.com/13495631/83800939-2ab98180-a6a0-11ea-906c-d393c65aedd3.png)

**Comment by jzabroski** [comment]
I think somewhere in the backlog I started spec'ing out something for restoring data as part of migrations, but I can't even recall my thought process at the time.  I think it involved copying all the related data to a separate schema and transfering it back, which would be a minimal t-log operation in case of rollback required.
